Minor clarifications of device specification (#152)
harp-tech/protocol: Device.md

## Device Interface

All device functionality provided to the host under the Harp protocol, including device configuration, is specified using a standardized device interface. The device interface is composed of a collection of hardware registers. Each register is defined by a unique memory location, which can be accessed for reading and writing, and potentially provide access to hardware-related functionality.

For example, a register can be used to configure device properties such as sampling frequency, operation mode, input / output routing, and other device-specific parameters. A register can also be used to control specific functionality, such as toggling digital output lines, starting and stopping pulse trains, or moving a motor.

Each register in the device interface is assigned a unique zero-based address, and a payload type describing the format of the data stored in the register. Registers can be read-only or allow both reading and writing.

> Except for [Optional or Deprecated registers](#optional-or-deprecated-registers), whose behavior is clarified below, all registers and functionality specified in this document MUST be implemented for a device to be declared compliant with that interface.

Author: Ye I figured when reading it. What I wanted to convey here is the ABSOLUTE necessity of devices implementing the core spec to be considered a valid "harp device". This is a non-negotiable in my mind. 

Past that, we can be more opinionated about the behavior of the device-specific interface, but the issue i see is that the device.yml not yet mentioned in the spec. Maybe something we should reconsider when tackling #124? For now, I would err on the side of scoping this document to the core registers.

I am also not sure if 
> all registers and functionality specified in the device interface MUST be implemented for a device to be declared compliant with that interface.

adds much. We would essentially be saying that all registers specified in a device.yml are expected to exist, and their functionality be implemented, in the device (not sure what the alternative would be? Sounds like a bug/bad design(?)). Since this is something left to developers I am not sure if there is much we could do to track/enforce this anyway.

> Except for [Optional or Deprecated registers](#optional-or-deprecated-registers), whose behavior is clarified below, all registers and functionality specified in this document MUST be implemented for a device to be declared compliant with that interface.

Reviewer: > the device.yml not yet mentioned in the spec

I agree we should definitely mention it, that was my intention to complete the section on "Device Interface" but it just gets so tangled that I couldn't find a way to keep things both simple and complete.

In my mind for a device to be a Harp device it really should both implement the core registers and provide a device.yml, otherwise it gets crazy.

But sure maybe we don't need to go yet so far as saying that they MUST do that to be a Harp device, in which case we can scale back the wording for now.

> Except for [Optional or Deprecated registers](#optional-or-deprecated-registers), whose behavior is clarified below, all registers and functionality specified in this document MUST be implemented for a device to be declared compliant with that interface.

Reviewer: ```suggestion
Except for [Optional or Deprecated registers](#optional-or-deprecated-registers), whose behavior is clarified below, all registers and functionality specified in this document MUST be implemented for a device to be declared compliant with this specification.
```

I agree, let's keep it simple for now. I just added in that case for consistency a change in the reference at the end of the document.

## Application Registers

All registers pertaining to the hardware-specific operation of a device are application registers. Every application register MUST have an address equal to, or greater than, 32. The address, payload type, and naming of application registers is left to the device developer, as it will always be specific to each device.

## Optional and Deprecated Registers

> Some registers are marked as **Optional** or **Deprecated** in the [Core Registers](#core-registers) table, and their functionality MAY not be fully implemented in all devices. However, all optional or deprecated registers MUST implement `Read` commands by sending a reply to the host with the specified default value. Moreover, they MUST be included as part of the [`R_OPERATION_CTRL`](#r_operation_ctrl-u8--operation-mode-configuration) register dump.

Author: ```suggestion
Some registers are marked as **Optional** or **Deprecated** in the table below ([Core Registers](#core-registers)), and their functionality MAY not be fully implemented in all devices. However, all optional or deprecated registers MUST implement `Read` commands by sending a reply to the host with the specified default value. Moreover, they MUST be included as part of the [`R_OPERATION_CTRL`](#r_operation_ctrl-u8--operation-mode-configuration) register dump.
```

How about this? I really just wanted to leave some anchoring on the document to make it explicit what/where the table is.

> Some registers are marked as **Optional** or **Deprecated** in the [Core Registers](#core-registers) table, and their functionality MAY not be fully implemented in all devices. However, all optional or deprecated registers MUST implement `Read` commands by sending a reply to the host with the specified default value. Moreover, they MUST be included as part of the [`R_OPERATION_CTRL`](#r_operation_ctrl-u8--operation-mode-configuration) register dump.

Reviewer: Ok, sounds good.

> Some registers are marked as **Optional** or **Deprecated** in the table below ([Core Registers](#core-registers)), and their functionality MAY not be fully implemented in all devices. However, all optional or deprecated registers MUST implement `Read` commands by sending a reply to the host with the specified default value. Moreover, they MUST be included as part of the [`R_OPERATION_CTRL`](#r_operation_ctrl-u8--operation-mode-configuration) register dump.

Reviewer: ```suggestion
Some registers are marked as **Optional** or **Deprecated** in the [Core Registers](#core-registers) table, and their functionality MAY not be fully implemented in all devices. However, all optional or deprecated registers MUST implement `Read` commands by sending a reply to the host with the specified default value. Moreover, they MUST be included as part of the [`R_OPERATION_CTRL`](#r_operation_ctrl-u8--operation-mode-configuration) register dump.
```

In the end I feel it might be better to just refer directly to the table without a parentheses.

    0      :d1, 0, 4
```

Contains the current system timestamp in whole seconds. The default value is `0` (Zero) and will increment one unit for each elapsed second.

### **`R_TIMESTAMP_MICRO` (U16) – System timestamp (microseconds)**

Address: `009`

> Contains the microseconds count within each second. Each LSB corresponds to 32 microseconds. The maximum value is 31249. The default value is `0` (Zero).

Author: Ye fixed. I moved it down to be consistent with all the other registers.

> [!WARNING]
>
> `Speed` mode is now deprecated, and SHOULD NOT be implemented in any new devices. This mode was used to support the implementation of dedicated communication protocols. When entering `Speed` mode, the device SHOULD NOT reply to host commands, other than its specific `Speed` mode commands.

> > 

Author: I have been playing with [pymarkdownlnt](https://pypi.org/project/pymarkdownlnt/) but some of the rules are a bit weird. We could try to settle on a config file we are happy with.

> > 

Reviewer: We can maybe try [something like this](https://github.com/vinibrsl/editorconfig-templates/blob/master/Markdown.editorconfig).

The basic test would be to add it as `.editorconfig` at the root of the repo and check if pymarkdownInt and VS code respect it? I feel the `trim_trailing_whitespace = false` might be enough, but the only way is trying.

I'm not sure we need `end_of_line = lf` though, since the version control system should take care of that.

For now I don't think we need to do anything in this PR, I can debug this on my end.

    0      :d1, 0, 1
```

> An array of 8 bytes that can be used to store a tag for a specific firmware build. For instance, it could be used to store the Git hash of a specific release/commit. The byte-order is little-endian. This register is read-only.

Reviewer: ```suggestion
An array of 8 bytes that can be used to store firmware-specific metadata. For instance, it could be used to store the Git hash of a specific release/commit. The byte-order is little-endian. This register is read-only.
```

Is this a "build" id, or a Git commit hash? The build ID is usually associated with knowing exactly when and where the build was generated, regardless of the source-code version but I don't think this is what we mean here.

"Version" would work technically for a commit hash in a version-control system, but I understand the confusion with the notion of official "release", so I tried to reword it to avoid clashing.

## Deprecated Core Registers

> The following registers are deprecated and their functionality SHOULD NOT be implemented in new devices. They MUST still exist as read-only registers, and included in the [`R_OPERATION_CTRL`](#r_operation_ctrl-u8--operation-mode-configuration) register dump. They are kept for backward compatibility with older devices and may be removed in future protocol versions.

Reviewer: ```suggestion
The following registers are deprecated and their functionality SHOULD NOT be implemented in new devices. They MUST still exist as read-only registers, and included in the [`R_OPERATION_CTRL`](#r_operation_ctrl-u8--operation-mode-configuration) register dump. They are kept for backward compatibility with older hosts and may be removed in future protocol versions.
```

Minor tweak: I think the main backward compatibility worry here is with older "hosts" correct?

Also just a note to say I am happy to recommend these registers not to be implemented at all, but in that case we should probably adjust the wording of the section at the top of the document for consistency, as these directives overlap.

In general I would prefer to avoid overlap in specs, so that we have only one place prescribing guidelines for each concept, in which case we can either remove the wording on deprecated registers from above, or merge both and remove the one here.
